### DISCUSSION TITLE
Use Upload scalar in input model

### DISCUSSION THREAD

**Discussion by clucutar-codefever** [graphql-java-kickstart/graphql-java-tools#498] [Q&A]
Hi,

I have the following mutation:
```
createContentFile(input: [ContentFileInput]): ContentFile

input ContentFileInput{
    path: String!
    file: Upload
}
```

From Postman I'm sending this input:
```
operations: {"operationName":"createContentFile","variables":{"input": [{"file": null, "path":"/cristian/Upload Test 2020-10-14 at 11"},{"file": null, "path":"/cristian/Upload Test 2020-10-14 at 11"}]},"query":"mutation createContentFile($input: [ContentFileInput]) {\n  createContentFile(input: $input) {\n    id\n    __typename\n  }\n}\n"}

map: { "file0": ["variables.input.0.file"] , "file1":["variables.input.1.file"]}
```

The purpose is to map property file to an object of type Part.
When executing the request, it throws an exception because it can't deserialize the Part object.

At the same time, the following mutation and input work fine:
```
createContentFile(input: [ContentFileInput], files: [Upload]): ContentFile

{"operationName":"createContentFile","variables":{"files": [null, null], "input": [{"path":"/cristian/Upload Test 2020-10-14 at 11"},{"path":"/cristian/Upload Test 2020-10-14 at 11"}]},"query":"mutation createContentFile($input: [ContentFileInput], $files: [Upload]) {\n  createContentFile(input: $input, files: $files) {\n    id\n    __typename\n  }\n}\n"}

{ "file0": ["variables.files.0"] , "file1":["variables.files.1"]}
```

I looked into the java and kt code in order to understand why when using the Upload scalar as a direct variable it works and why it doesn't when the scalar is in the input object. The difference seem to be that in the first case, the object of type Part is just passed on while in the second case is deserialized, leading to an exception.

L.E. It seems that the problem comes from the fact that class MethodFieldResolver.kt detects scalar types only when they are endpoint parameters. If a scalar is part of a broader input model, it is seen as a JSON and sent to ObjectMapper for deserialization. Deserializing an object of type Part is not possible.

Any help is appreciated.

Regards,
Cristian.

**Comment by oryan-block** [comment]
Hi Cristian, unfortunately this is a [known limitation](https://github.com/graphql-java-kickstart/graphql-java-tools/issues/466).

One way to overcome this is by replacing `Part` with `String` in your resolvers while holding a reference to the actual files in your context object.
In your controller method, before calling the graphql executor, you'd want to update the query with file references: `"input": [{"file": "file0"...,{"file": "file1"...`. This is the tricky part since you'd need to parse the query json to get to the specific nodes that represent files.
Then you'd need to set a mapping from these references to the file parts in your context object `Map<String, Part/MultipartFile> files`
And you should be able to retrieve the file parts from the context:
```
    public void createContentFile(
        List<ContentFileInput> input,
        GraphQLContext context
    ) {
        context.getFile(input.get(0).getFile());
        ...
    }

    class ContentFileInput {
        String file;
        String path;
    }
```

---

Another workaround is to help jackson serialize the `Part` objects and it's [described here](https://gist.github.com/oryan-block/7630eb6d352778164c7fbdbb45e24265). 

I hope this helps.
